File: sdk/reproit-backend-go/instrument_test.go
package reproitbackend

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// tracedRequest runs one round trip through the Transport with an ambient
// capture-mode trace and returns the trace's recorded events.
func tracedRequest(t *testing.T, server *httptest.Server, body string) *BackendTrace {
	t.Helper()
	trace, err := Begin(&TraceContext{TraceID: "cap-x-1", CaptureEnvelope: true}, "GET /quote",
		BeginOptions{Input: HTTPInput{Query: map[string]any{"symbol": "ACME"}}.Value()})
	if err != nil {
		t.Fatal(err)
	}
	client := WrapClient(server.Client())
	method := http.MethodGet
	var reader io.Reader
	if body != "" {
		method = http.MethodPost
		reader = strings.NewReader(body)
	}
	request, err := http.NewRequest(method, server.URL+"/prices?tier=gold", reader)
	if err != nil {
		t.Fatal(err)
	}
	if body != "" {
		request.Header.Set("Content-Type", "application/json")
	}
	request = request.WithContext(ContextWithTrace(request.Context(), trace))
	response, err := client.Do(request)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := io.ReadAll(response.Body); err != nil {
		t.Fatal(err)
	}
	if err := response.Body.Close(); err != nil {
		t.Fatal(err)
	}
	return trace
}

func exchangeOf(t *testing.T, trace *BackendTrace) map[string]any {
	t.Helper()
	for _, event := range trace.Events() {
		if exchange, ok := event["exchange"].(map[string]any); ok {
			return exchange
		}
	}
	t.Fatal("no exchange recorded on the trace")
	return nil
}

func TestTransportRecordsRequestAndResponse(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(
		func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "application/json")
			_, _ = w.Write([]byte(`{"prices":[1,2],"apiKey":"sk-live-secret"}`))
		}))
	defer server.Close()

	exchange := exchangeOf(t, tracedRequest(t, server, `{"amount":5,"password":"hunter22"}`))
	if exchange["protocol"] != "http" {
		t.Fatalf("protocol wrong: %v", exchange)
	}
	request := exchange["request"].(map[string]any)
	if request["method"] != http.MethodPost {
		t.Fatalf("request method wrong: %v", request)
	}
	requestBody := request["body"].(map[string]any)
	if number, _ := requestBody["amount"].(json.Number); number.String() != "5" {
		t.Fatalf("request body lost: %v", requestBody)
	}
	// Structural redaction applies INSIDE captured exchange bodies.
	if secret, ok := requestBody["password"].(map[string]any); !ok ||
		secret["$reproit"] == nil {
		t.Fatalf("request body secret not redacted: %v", requestBody)
	}
	response := exchange["response"].(map[string]any)
	if status, _ := response["status"].(json.Number); status.String() != "200" {
		t.Fatalf("response status wrong: %v", response)
	}
	responseBody := response["body"].(map[string]any)
	prices := responseBody["prices"].([]any)
	if len(prices) != 2 {
		t.Fatalf("response body lost: %v", responseBody)
	}
	if secret, ok := responseBody["apiKey"].(map[string]any); !ok ||
		secret["$reproit"] == nil {
		t.Fatalf("response body secret not redacted: %v", responseBody)
	}
	if _, ok := response["headers"].(map[string]any); !ok {
		t.Fatalf("response headers not recorded: %v", response)
	}
}

func TestOversizedBodyKeepsProvableIdentityOnly(t *testing.T) {
	oversized := strings.Repeat("x", MaxExchangeBodyBytes+1)
	server := httptest.NewServer(http.HandlerFunc(
		func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Content-Type", "text/plain")
			_, _ = w.Write([]byte(oversized))
		}))
	defer server.Close()

	exchange := exchangeOf(t, tracedRequest(t, server, ""))
	response := exchange["response"].(map[string]any)
	if truncated, _ := response["truncated"].(bool); !truncated {
		t.Fatalf("oversized body not marked truncated: %v", response)
	}
	if _, present := response["body"]; present {
		t.Fatalf("oversized body retained bytes: %v", response)
	}
	if bytesRecorded, _ := response["bodyBytes"].(json.Number); bytesRecorded.String() !=
		"8193" {
		t.Fatalf("oversized body byte count wrong: %v", response)
	}
	digest := sha256.Sum256([]byte(oversized))
	if response["bodySha256"] != hex.EncodeToString(digest[:]) {
		t.Fatalf("oversized body digest wrong: %v", response)
	}
}

func TestTransportWithoutAmbientTraceRecordsNothing(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(
		func(w http.ResponseWriter, r *http.Request) {
			_, _ = w.Write([]byte("{}"))
		}))
	defer server.Close()

	trace, err := Begin(&TraceContext{TraceID: "cap-x-2"}, "GET /quote", BeginOptions{})
	if err != nil {
		t.Fatal(err)
	}
	// No ContextWithTrace: the call is outside any traced request.
	response, err := WrapClient(server.Client()).Get(server.URL + "/prices")
	if err != nil {
		t.Fatal(err)
	}
	_, _ = io.ReadAll(response.Body)
	_ = response.Body.Close()
	for _, event := range trace.Events() {
		if _, ok := event["exchange"]; ok {
			t.Fatal("an untraced call recorded an exchange")
		}
	}
}

func TestRunDBRecordsRowsAndErrors(t *testing.T) {
	trace, err := Begin(&TraceContext{TraceID: "cap-x-3", CaptureEnvelope: true},
		"GET /quote", BeginOptions{})
	if err != nil {
		t.Fatal(err)
	}
	ctx := ContextWithTrace(t.Context(), trace)
	_, err = RunDB(ctx, "SELECT id FROM issuers WHERE symbol = $1", []any{"ACME"},
		func() (DBOutcome, error) {
			return DBOutcome{
				Command: "SELECT", RowCount: 1,
				Rows: []any{map[string]any{"id": 7}},
			}, nil
		})
	if err != nil {
		t.Fatal(err)
	}
	_, err = RunDB(ctx, "SELECT boom", nil, func() (DBOutcome, error) {
		return DBOutcome{}, &DBError{Message: "relation missing", Code: "42P01"}
	})
	if err == nil {
		t.Fatal("live db error was swallowed")
	}
	exchanges := []map[string]any{}
	for _, event := range trace.Events() {
		if exchange, ok := event["exchange"].(map[string]any); ok {
			exchanges = append(exchanges, exchange)
		}
	}
	if len(exchanges) != 2 {
		t.Fatalf("expected two db exchanges, got %d", len(exchanges))
	}
	request := exchanges[0]["request"].(map[string]any)
	if request["values"].([]any)[0] != "ACME" {
		t.Fatalf("statement values lost: %v", request)
	}
	rows := exchanges[0]["response"].(map[string]any)["rows"].([]any)
	if len(rows) != 1 {
		t.Fatalf("recorded rows lost: %v", exchanges[0]["response"])
	}
	failure := exchanges[1]["response"].(map[string]any)["error"].(map[string]any)
	if failure["message"] != "relation missing" || failure["code"] != "42P01" {
		t.Fatalf("recorded db error lost: %v", failure)
	}
}

func TestCaptureModeStampsTheDeterminismEnvelope(t *testing.T) {
	trace, err := Begin(&TraceContext{TraceID: "cap-x-4", CaptureEnvelope: true},
		"GET /quote", BeginOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if _, ok := trace.Events()[0]["at"].(json.Number); !ok {
		t.Fatalf("capture-mode event lacks a wall-clock stamp: %v", trace.Events()[0])
	}
	if _, ok := trace.Events()[0]["monoNs"].(json.Number); !ok {
		t.Fatalf("capture-mode event lacks a monotonic stamp: %v", trace.Events()[0])
	}
	// Scan-time traces must stay byte-stable: no envelope fields at all.
	scan, err := Begin(&TraceContext{TraceID: "trace-a"}, "GET /quote", BeginOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if _, present := scan.Events()[0]["at"]; present {
		t.Fatalf("scan-time event gained an envelope stamp: %v", scan.Events()[0])
	}
	if _, present := scan.Events()[0]["monoNs"]; present {
		t.Fatalf("scan-time event gained an envelope stamp: %v", scan.Events()[0])
	}
}

func TestCapabilitiesClaimNetworkOnlyWithExchanges(t *testing.T) {
	plain := captureCapabilities(capturedOperation{events: []map[string]any{
		{"kind": "effect", "effect": "write"},
	}})
	for _, item := range plain {
		if item.(map[string]any)["capability"] == "network" {
			t.Fatal("network capability claimed without any recorded exchange")
		}
	}
	withExchange := captureCapabilities(capturedOperation{events: []map[string]any{
		{"kind": "effect", "exchange": map[string]any{"protocol": "http"}},
	}})
	found := false
	for _, item := range withExchange {
		if item.(map[string]any)["capability"] == "network" {
			found = true
		}
	}
	if !found {
		t.Fatal("network capability missing despite a recorded exchange")
	}
}

func TestResolveCommitPrefersConfigThenEnvironment(t *testing.T) {
	if got := resolveCommit("abc123"); got != "abc123" {
		t.Fatalf("configured commit ignored: %q", got)
	}
	t.Setenv("REPROIT_COMMIT", "env-commit")
	if got := resolveCommit(""); got != "env-commit" {
		t.Fatalf("REPROIT_COMMIT ignored: %q", got)
	}
	t.Setenv("REPROIT_COMMIT", "not a token")
	t.Setenv("GITHUB_SHA", "sha-fallback")
	if got := resolveCommit(""); got != "sha-fallback" {
		t.Fatalf("GITHUB_SHA fallback ignored: %q", got)
	}
}

File: sdk/reproit-backend-php/test/instrument_test.php
<?php

/*!
 * Outbound-exchange capture and hermetic replay tests for the PHP SDK.
 *
 * Run: php test/instrument_test.php
 */

declare(strict_types=1);

namespace ReproitBackend\Test;

use ReproitBackend\BackendTrace;
use ReproitBackend\Capture;
use ReproitBackend\Instrument;

use function ReproitBackend\bounded_body;
use function ReproitBackend\bounded_headers;
use function ReproitBackend\http_exchange;

use const ReproitBackend\MAX_EXCHANGE_BODY_BYTES;
use const ReproitBackend\MAX_EXCHANGE_HEADERS;

require_once __DIR__ . '/support.php';
require_once __DIR__ . '/../reproit.php';

/** Streams preset chunks one per read: the shape an SSE upstream delivers. */
final class ChunkStream implements \Psr\Http\Message\StreamInterface
{
    private int $at = 0;

    /** @param list<string> $chunks */
    public function __construct(private readonly array $chunks)
    {
    }

    public function read(int $length): string
    {
        if ($this->at >= \count($this->chunks)) {
            return '';
        }
        $chunk = $this->chunks[$this->at];
        $this->at += 1;
        return $chunk;
    }

    public function eof(): bool
    {
        return $this->at >= \count($this->chunks);
    }

    public function isSeekable(): bool
    {
        return true;
    }

    public function rewind(): void
    {
        $this->at = 0;
    }

    public function getContents(): string
    {
        $rest = implode('', \array_slice($this->chunks, $this->at));
        $this->at = \count($this->chunks);
        return $rest;
    }

    public function close(): void
    {
    }

    public function detach()
    {
        return null;
    }

    public function getSize(): ?int
    {
        return \strlen(implode('', $this->chunks));
    }

    public function tell(): int
    {
        return 0;
    }

    public function getMetadata(?string $key = null)
    {
        return null;
    }
}

/** Just enough PSR-7 response for the decorator's capture tee. */
final class StubResponse implements \Psr\Http\Message\ResponseInterface
{
    public function __construct(
        private readonly int $status,
        private readonly array $headers,
        private object $body,
    ) {
    }

    public function getStatusCode(): int
    {
        return $this->status;
    }

    public function getHeaders(): array
    {
        return array_map(fn (string $value): array => [$value], $this->headers);
    }

    public function getHeaderLine(string $name): string
    {
        return $this->headers[strtolower($name)] ?? '';
    }

    public function getBody(): object
    {
        return $this->body;
    }

    public function withBody(object $body): self
    {
        $clone = clone $this;
        $clone->body = $body;
        return $clone;
    }
}

/** Inner PSR-18 client answering with a preset response. */
final class StubClient implements \Psr\Http\Client\ClientInterface
{
    public function __construct(private readonly StubResponse $response)
    {
    }

    public function sendRequest(
        \Psr\Http\Message\RequestInterface $request,
    ): \Psr\Http\Message\ResponseInterface {
        return $this->response;
    }
}

/** Just enough PSR-7 request for the decorator: method, URI, body, headers. */
final class StubRequest implements \Psr\Http\Message\RequestInterface
{
    public function __construct(
        private readonly string $method,
        private readonly string $url,
        private readonly string $body = '',
        private readonly array $headers = [],
    ) {
    }

    public function getMethod(): string
    {
        return $this->method;
    }

    public function getUri(): StubUri
    {
        return new StubUri($this->url);
    }

    public function getHeaders(): array
    {
        return array_map(fn (string $value): array => [$value], $this->headers);
    }

    public function getHeaderLine(string $name): string
    {
        return $this->headers[strtolower($name)] ?? '';
    }

    public function getBody(): ChunkStream
    {
        return new ChunkStream($this->body === '' ? [] : [$this->body]);
    }
}

final class StubUri
{
    public function __construct(private readonly string $url)
    {
    }

    public function getHost(): string
    {
        return (string) (parse_url($this->url, PHP_URL_HOST) ?: '');
    }

    public function __toString(): string
    {
        return $this->url;
    }
}

function trace_with_envelope(): BackendTrace
{
    return BackendTrace::begin(
        [
            'traceId' => 'cap-x-1',
            'actor' => null,
            'actionIndex' => 0,
            'build' => null,
            'configContract' => null,
            'captureEnvelope' => true,
        ],
        'GET /quote',
        ['input' => ['query' => ['symbol' => 'ACME']]]
    );
}

/** @return list<array> */
function recorded_exchanges(BackendTrace $trace): array
{
    $found = [];
    foreach ($trace->events() as $event) {
        if (\is_array($event['exchange'] ?? null)) {
            $found[] = $event['exchange'];
        }
    }
    return $found;
}

// bounds: an oversized body keeps provable identity only
$big = str_repeat('x', MAX_EXCHANGE_BODY_BYTES + 1);
$bounded = bounded_body($big, 'text/plain');
check_same(true, $bounded['truncated'], 'oversized body is marked truncated');
check_same(\strlen($big), $bounded['bodyBytes'], 'oversized body keeps its byte count');
check(
    preg_match('/^[0-9a-f]{64}$/', $bounded['bodySha256']) === 1,
    'oversized body keeps a sha256'
);
check(!isset($bounded['body']), 'oversized body drops its content');

// bounds: headers are capped
$many = [];
for ($index = 0; $index < MAX_EXCHANGE_HEADERS + 10; $index++) {
    $many['x-header-' . $index] = (string) $index;
}
check_same(
    MAX_EXCHANGE_HEADERS,
    \count(bounded_headers($many)['headers']),
    'headers are capped at the shared bound'
);

// redaction applies INSIDE captured exchange bodies
$trace = trace_with_envelope();
Instrument::setTrace($trace);
Instrument::record('call', 'pricing', 'GET /prices', http_exchange(
    ['method' => 'GET', 'url' => 'http://pricing/prices', 'headers' => [], 'body' => null,
     'contentType' => ''],
    [
        'status' => 200,
        'headers' => ['content-type' => 'application/json'],
        'body' => json_encode(['prices' => [1, 2], 'apiKey' => 'sk-live-secret']),
        'contentType' => 'application/json',
    ]
));
$exchange = recorded_exchanges($trace)[0] ?? null;
check(\is_array($exchange), 'exchange recorded on the ambient trace');
check_same(200, $exchange['response']['status'], 'response status recorded');
check_same([1, 2], $exchange['response']['body']['prices'], 'response body recorded');
check_same(
    true,
    $exchange['response']['body']['apiKey']['$reproit']['redacted'],
    'secrets inside exchange bodies are redacted at source'
);

// the database boundary records rows and errors
$trace = trace_with_envelope();
Instrument::setTrace($trace);
Instrument::db('SELECT id FROM issuers WHERE symbol = $1', ['ACME'], fn () => [
    'command' => 'SELECT', 'rowCount' => 1, 'rows' => [['id' => 7]],
]);
try {
    Instrument::db('SELECT boom', null, function (): array {
        throw new \RuntimeException('relation missing');
    });
} catch (\RuntimeException) {
    // recorded below
}
$exchanges = recorded_exchanges($trace);
check_same(2, \count($exchanges), 'both database statements recorded');
check_same('db', $exchanges[0]['protocol'], 'database protocol tagged');
check_same(['ACME'], $exchanges[0]['request']['values'], 'bound values recorded');
check_same([['id' => 7]], $exchanges[0]['response']['rows'], 'rows recorded');
check_same(
    'relation missing',
    $exchanges[1]['response']['error']['message'],
    'a driver error is recorded, not swallowed'
);
Instrument::setTrace(null);

// PSR-18 decoration: the response body is TEED, not drained. Chunk
// boundaries are recorded exactly as the app consumed them (the SSE / LLM
// streaming shape), the request line and body ride the exchange, and the
// record lands at the moment the app observes EOF.
$trace = trace_with_envelope();
Instrument::setTrace($trace);
$sse = new StubResponse(
    200,
    ['content-type' => 'text/event-stream'],
    new ChunkStream(["data: a\n\n", "data: b\n\n", "data: c\n\n"])
);
$client = new \ReproitBackend\Psr18\RecordingClient(new StubClient($sse));
$request = new StubRequest(
    'POST',
    'http://llm.internal/v1/stream',
    '{"model":"m","apiKey":"sk-live-leak"}',
    ['content-type' => 'application/json']
);
$teed = $client->sendRequest($request);
check_same([], recorded_exchanges($trace), 'nothing recorded before the app reads');
$seen = [];
$body = $teed->getBody();
while (!$body->eof()) {
    $chunk = $body->read(65536);
    if ($chunk !== '') {
        $seen[] = $chunk;
    }
}
check_same(
    ["data: a\n\n", "data: b\n\n", "data: c\n\n"],
    $seen,
    'the app reads the live stream through the tee untouched'
);
$exchange = recorded_exchanges($trace)[0] ?? null;
check(\is_array($exchange), 'the exchange records at EOF');
check_same('http', $exchange['protocol'] ?? null, 'psr18 exchange protocol');
check_same('POST', $exchange['request']['method'], 'request method recorded');
check_same('http://llm.internal/v1/stream', $exchange['request']['url'], 'url recorded');
check_same(
    true,
    $exchange['request']['body']['apiKey']['$reproit']['redacted'] ?? null,
    'request body secrets redacted at source'
);
check_same("data: a\n\ndata: b\n\ndata: c\n\n", $exchange['response']['body'], 'body verbatim');
check_same([9, 9, 9], $exchange['response']['stream']['chunks'], 'observed chunk boundaries');

// An abandoned body records nothing, exactly like a response nobody reads.
$before = \count(recorded_exchanges($trace));
$client->sendRequest(new StubRequest('GET', 'http://llm.internal/ignored'));
check_same(
    $before,
    \count(recorded_exchanges($trace)),
    'an abandoned response body records no exchange'
);
Instrument::setTrace(null);

// PDO wrap: statements record the Node pg wire shape and the app observes
// exactly the recorded rows. Guarded on the sqlite driver so the suite stays
// dependency-free; the skip is loud, never silent.
if (\in_array('sqlite', \PDO::getAvailableDrivers(), true)) {
    $trace = trace_with_envelope();
    $pdo = new \ReproitBackend\RecordingPdo('sqlite::memory:');
    $pdo->exec('CREATE TABLE issuers (id INTEGER PRIMARY KEY, symbol TEXT)');
    $pdo->exec("INSERT INTO issuers (id, symbol) VALUES (7, 'ACME')");
    Instrument::setTrace($trace);
    $statement = $pdo->prepare('SELECT id, symbol FROM issuers WHERE symbol = ?');
    $statement->execute(['ACME']);
    $rows = $statement->fetchAll();
    $failed = false;
    try {
        $pdo->query('SELECT boom FROM nowhere');
    } catch (\PDOException) {
        $failed = true;
    }
    Instrument::setTrace(null);
    check_same([['id' => 7, 'symbol' => 'ACME']], $rows, 'pdo app rows are the driver rows');
    $exchanges = recorded_exchanges($trace);
    check_same(2, \count($exchanges), 'both pdo statements recorded');
    check_same('pg', $exchanges[0]['protocol'], 'pdo exchanges use the pg wire shape');
    check_same(
        'SELECT id, symbol FROM issuers WHERE symbol = ?',
        $exchanges[0]['request']['text'],
        'statement text recorded'
    );
    check_same(['ACME'], $exchanges[0]['request']['values'], 'bound values recorded');
    check_same('SELECT', $exchanges[0]['response']['command'], 'command tag recorded');
    check_same(
        [['id' => 7, 'symbol' => 'ACME']],
        $exchanges[0]['response']['rows'],
        'recorded rows equal the rows the app saw'
    );
    check($failed, 'a driver error still raises to the app');
    check(
        \is_string($exchanges[1]['response']['error']['message'] ?? null),
        'the driver error is recorded, not swallowed'
    );
} else {
    fwrite(STDERR, "SKIP: pdo_sqlite driver unavailable; PDO capture cases not run\n");
}

// capture mode stamps the envelope; scan-time traces stay byte-stable
$stamped = trace_with_envelope();
$allStamped = true;
foreach ($stamped->events() as $event) {
    if (!\is_int($event['at'] ?? null) || !\is_int($event['monoNs'] ?? null)) {
        $allStamped = false;
    }
}
check($allStamped, 'capture-mode events carry wall-clock and monotonic stamps');
$scan = BackendTrace::begin(
    ['traceId' => 'trace-a', 'actor' => null, 'actionIndex' => 0, 'build' => null,
     'configContract' => null],
    'op'
);
$scanClean = true;
foreach ($scan->events() as $event) {
    if (\array_key_exists('at', $event) || \array_key_exists('monoNs', $event)) {
        $scanClean = false;
    }
}
check($scanClean, 'scan-time events carry no envelope stamps');

// the batch declares `network` only when exchanges exist
$capture = Capture::create([
    'endpoint' => 'http://c/v1/capture-batches', 'apiKey' => 'sk', 'appId' => 'app-demo',
]);
$withExchange = BackendTrace::begin($capture->context(), 'GET /quote', ['input' => null]);
$withExchange->effect('call', [
    'resource' => 'pricing',
    'key' => 'GET /prices',
    'exchange' => [
        'protocol' => 'http',
        'request' => ['method' => 'GET', 'url' => 'http://pricing/prices'],
        'response' => ['status' => 200, 'body' => ['prices' => null]],
    ],
]);
$withExchange->finish(['error' => 'boom'], 500, false, true);
$batch = $capture->buildBatch([
    ['operation' => 'GET /quote', 'status' => 500, 'events' => $withExchange->events()],
]);
$names = array_map(fn (array $entry) => $entry['capability'], $batch['capabilities']);
check(\in_array('network', $names, true), 'network capability declared with exchanges');

$plain = BackendTrace::begin($capture->context(), 'GET /quote', ['input' => null]);
$plain->effect('read', ['resource' => 'inventory', 'key' => 'widget']);
$plain->finish(null, 500, false, true);
$bare = $capture->buildBatch([
    ['operation' => 'GET /quote', 'status' => 500, 'events' => $plain->events()],
]);
$bareNames = array_map(fn (array $entry) => $entry['capability'], $bare['capabilities']);
check(
    !\in_array('network', $bareNames, true),
    'no network capability without recorded exchanges'
);

report('instrument_test');
